### ISSUE TITLE
id collision

### ISSUE THREAD

**Issue by ozum** [vuex-orm/vuex-orm#261]
I mentioned this in a PR thread, but this is an important issue and should be in its own case.

I see that composite primary keys (ids) are merged using `_`. However it may cause collision in example below:

I didn't encounter this situation, because my keys are UUID, but just wanted to bring it in attention.

```js
export default class Option extends Model {
  static entity = "options";

  static primaryKey = ["id", "otherId"];

  static fields() {
    return {
      id: this.string(""),
      otherId: this.string(""),
    };
  }
}

Option.create({
  data: [
    { id: "a_", otherId: "b" }, // COLLISION: $id: "a__b"
    { id: "a", otherId: "_b" }, // COLLISION: $id: "a__b"
  ]
})
```

It may be better to store composite keys in an array: `$id: ["a_", "b"]`.

**Comment by kiaking** [comment]
Ahhh... Yeah that's true. Good point, thanks for the feedback!

Well we can't store composite keys as array because why we're merging the keys is to use it as the key of the records.

```js
{
  posts: {
    data: {
      `a_b`: { ... } // <- The key have to be a single string.
    }
  }
}
```

But yes, we should avoid such collision... Maybe prefixing with `_` or maybe double underscore...? So it will be something like `__a_b`?

**Comment by ozum** [comment]
I think `JSON.stringify()` is best bet then. Generates strings from arrays relatively fast and without collision risk. Also you can convert it back to array using `JSON.parse()` in case it is needed.

```js
JSON.stringify(['firstKey', 'secondKey']);  // --> ["firstKey","secondKey"]
JSON.stringify(['"firstKey"', '"secondKey"']);  // --> ["\"firstKey\"","\"secondKey\""]
```

**Comment by kiaking** [comment]
!!! Yeah your genius. Never thought of that approach. That should work definitely. Let's put this on the list for the future implementation.

**Comment by ozum** [comment]
By the way, how does `find` works with array keys (composite/compound keys)? I'm asking, because I can update `findIn`, `whereId` and `whereIdIn` according to this.

**Comment by kiaking** [comment]
Currently it does nothing special. It looks up for the key with specified string. Within the state, the composite keys are merged as a string like `a_b`, so if you specify that key name, it fetches the record.

```js
query.find('a_b')
```

It would be great to support this! Haven't though but something like following API looks nice...? Not sure.

```js
query.find({ key_1: 'a', key_2: 'b' })

query.findIn([{ key_1: 'a', key_2: 'b' }, { key_1: 'b', key_2: 'b' }])
```

**Comment by ozum** [comment]
Keys are already ordered in model:
```js
export default class Option extends Model {
  static primaryKey = ["key_1", "key_2", "key_3"];
}
```

So we don't need to feed key names.

```js
// Key 1 , Key 2, Key 3
query.find(['a', 'b', 'c'])

// Arrays of Key 1, Key 2, Key3
query.findIn([['a', 'b', 'c'], ['x', 'y', 'z']])
```

This was the reason I write separate `findIn` methods instead of putting it in `find`. Otherwise we could not decide whether `find([1,2])` was array id lookup or multiple records with single id.

It seems easy to implement then. It is just needed to update key lookups in related methods. Below is example for `find`

```js
// Function to generate key. You may want to change implementation in the future.
// Other methods will not need to be changed.
// This can be an instance or static class function.
getKey (id: number | string | Array<number | string>): string {
  return Array.isArray(id) ? JSON.stringify(id) : id
}

...

find (id: number | string | Array<number | string>): Data.Item {
  return this.item(this.state.data[this.model.getKey(id)]);
}

```

**Comment by kiaking** [comment]
This is now implemented with Vuex ORM 0.33.0! 🎉 
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.33.0
